fcrepo4/fcrepo4
# Allow Datetime negotiation on versioned resources

**JIRA Ticket**: https://jira.duraspace.org/browse/FCREPO-2613

# What does this Pull Request do?
Adds the ability for a versioned resource to take an `Accept-Datetime` header and return:
* a `302 Found` to the Memento which is the closest but older version (if more than one Memento)
* a `302 Found` to the Memento regardless of older or not (if only one Memento)
* a `404 Not Found` if there are no Mementos. 

# What's new?
Added `FedoraResource findMementoByDatetime(Instant mementoDatetime);` to `FedoraResource` interface and added an implementation to `FedoraResourceImpl`.

Added `getMemento()` to `FedoraLdp` if there is an `Accept-Datetime` header and the resource is versioned. This takes over the interaction and returns the headers for the normal resource but adds either a `302 Found` with a `Location` header to the Memento or a `404 Not Found` if there are no Mementos.

# How should this be tested?

1. Build and run it.
1. Create a non-versioned resource, send a GET with an `Accept-Datetime` header. You get the resource as normal
1. Make a versioned resource 
1. Send a GET request to the versioned resource with an `Accept-Datetime: 2018-04-17T01:01:01Z` header (ie. not RFC-1123) get a `400 Bad Request`.
1. Send a GET request to the versioned resource with an `Accept-Datetime: Tue, 17 Apr 2018 19:04:17 GMT` header (ie. is RFC-1123) get a `404 Not Found`.
1. Make a version of the resource
1. Send a GET request to the versioned resource with an `Accept-Datetime: Tue, 17 Apr 2018 19:04:17 GMT` header (ie. is RFC-1123) get a `302 Found` with a `Location` header to the Memento.

Make more Mementos and requests and ensure you always get the Memento with a datetime just older than your request.

Example:
* Does this change require documentation to be updated? Probably
* Does this change add any new dependencies? no
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? no
* Could this change impact execution of existing code? no

# Interested parties
@dbernstein @bseeger @bbpennel @yinlinchen
